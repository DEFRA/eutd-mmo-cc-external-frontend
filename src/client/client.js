// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
