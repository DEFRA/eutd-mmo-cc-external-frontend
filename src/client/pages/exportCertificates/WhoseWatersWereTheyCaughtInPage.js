import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Checkbox from '../../components/elements/checkbox';

import {
  Main,
  Header,
  GridRow,
  GridCol,
  InputField,
  MultiChoice
} from 'govuk-react';

import PageTitle from '../../components/PageTitle';

import {
  addConservation,
  saveConservation,
  dispatchClearErrors,
  dispatchApiCallFailed,
  getConservation,
  clearConservation,
  getExportPayload
} from '../../actions';

import {
  getLandingType
} from '../../actions/landingsType.actions';

import BackLinkWithErrorClearOut from '../../components/elements/BackLinkWithErrorClearOut';
import ErrorIsland from '../../components/elements/ErrorIsland';
import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import { scrollToErrorIsland, onHandleErrorClick } from '../utils';
import HelpLink from '../../components/HelpLink';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';
import BackToProgressLink from '../../components/BackToProgressLink';
import { withTranslation } from 'react-i18next';

class WhoseWatersWereTheyCaughtInPage extends Component {

  state = {
    jsEnabled: false
  };

  constructor(props) {
    super(props);
  }

  otherWatersChange = (e) => {
    this.props.addConservation({ [e.target.name]: e.target.value });
  };

  onChange = (e) => {
    this.props.addConservation({ [e.target.name]: (e.target.checked ? 'Y' : undefined) });
  };

  onSubmit = async (e) => {
    e.preventDefault();
    try {
      await this.doSubmit(false);
      const { nextUri } = this.props.route;
      const documentNumber = (this.props.match && this.props.match.params)
        ? this.props.match.params['documentNumber']
        : '';
      const nextUrl = nextUri.replace(':documentNumber', documentNumber);
      this.props.history.push(nextUrl);
    } catch (error) {
      scrollToErrorIsland();
    }
  };

   hasRequiredData() {
     return (this.props.exportPayload.items && this.props.exportPayload.items.length > 0);
  }

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    try {
      const { saveAsDraftUri } = this.props.route;
      await this.doSubmit(true);
      this.props.history.push(saveAsDraftUri);
    } catch (error) {
      scrollToErrorIsland();
    }
  }

  doSubmit = async (isConservationSavedAsDraft) => {
    const { path, nextUri } = this.props.route;
    const documentNumber = (this.props.match && this.props.match.params)
      ? this.props.match.params['documentNumber']
      : '';

    await this.props.saveConservation(path, nextUri, isConservationSavedAsDraft, documentNumber);
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  displayLandingEntryOptions() {
    const documentNumber = this.props.match.params['documentNumber'];
    const { landingsEntryUri } = this.props.route;
    this.props.history.push(landingsEntryUri.replace(':documentNumber', documentNumber));
  }

  hasLandingEntryOption() {
    const { getAddedLandingsType } = this.props;
    return getAddedLandingsType && getAddedLandingsType.landingsEntryOption !== null && getAddedLandingsType.generatedByContent === false;
  }

  componentWillUnmount() {
    this.props.clearConservation();
  }

  componentDidMount = async () => {
    const documentNumber = this.props.match.params['documentNumber'];

    try {
      await this.props.getConservation(documentNumber);
      await this.props.getLandingType(documentNumber);
      await this.props.getExportPayload(documentNumber);

      if (!this.hasLandingEntryOption()) {
        this.displayLandingEntryOptions();
      }
    } catch (e) {
      console.log(e);
    }
    this.setState({ jsEnabled: true });
  };

  componentDidUpdate() {
    const { conservation } = this.props;

    if (conservation.unauthorised === true) {
      this.unauthorised();
    }
  }

  render = () => {
    const { errors, otherWatersError, watersCaughtInError, } = this.props.errors;

    const { otherWaters, caughtInUKWaters, caughtInEUWaters, caughtInOtherWaters } = this.props.conservation;
    const { dispatchClearErrors: _dispatchClearErrors, history, route, t } = this.props;
    const { previousUri, path, nextUri, saveAsDraftUri, journey, directLandingUri, progressUri, title } = route;
    const saveAsDraftFormActionUrl = '/orchestration/api/v1/conservation/saveAsDraft';

    // include documentNumber within route object
    const documentNumber = (this.props.match && this.props.match.params)
      ? this.props.match.params['documentNumber']
      : '';

    const backTo =
      this.props.getAddedLandingsType.landingsEntryOption === 'directLanding'
        ? directLandingUri.replace(':documentNumber', documentNumber)
        : previousUri.replace(':documentNumber', documentNumber);

    return (
      <Main>
        <PageTitle title={`${errors ? `${t('commonPageTitleErrorText')} ` : ''}${t('ccWhoseWatersWereTheyCaughtInTitleText')} - ${t(title)}`} />
        {errors && <ErrorIsland errors={errors.map(err => ({ message: t(err.text), key: err.targetName }))} onHandleErrorClick={onHandleErrorClick} />}
        <GridRow>
          <GridCol>
            {this.hasRequiredData() &&
              (<BackLinkWithErrorClearOut
                backUri={backTo}
                clearErrors={_dispatchClearErrors}
                history={history}
                labeltext={t('commonBackLinkBackButtonLabel')}
              />)}
            <Header>{t('ccWhoseWatersWereTheyCaughtInHeaderText')}</Header>
          </GridCol>
        </GridRow>
        <form action="/orchestration/api/v1/conservation" method="POST" onSubmit={this.onSubmit}>
          <input type='hidden' name='currentUri' value={path} />
          <input type='hidden' name='nextUri' value={nextUri} />
          <input type="hidden" name="dashboardUri" value={saveAsDraftUri} />
          <input type="hidden" name="journey" value={journey} />
          <div style={{ display: 'none' }}>
            <Checkbox></Checkbox>
          </div>

          <MultiChoice meta={{ touched: t(watersCaughtInError), error: t(watersCaughtInError) }} label={t('ccWhoseWatersWereTheyCaughtInSelectText')}>
            <Checkbox id="watersCaughtIn" name="caughtInUKWaters" onChange={this.onChange} checked={caughtInUKWaters === 'Y'} value="Y">{t('ccWhoseWatersWereTheyCaughtInCheckBox1Text')}</Checkbox>
            <Checkbox id="caughtInEUWaters" name="caughtInEUWaters" onChange={this.onChange} checked={caughtInEUWaters === 'Y'} value="Y">{t('ccWhoseWatersWereTheyCaughtInCheckBox2Text')}</Checkbox>
            <Checkbox id="caughtInOtherWaters" name="caughtInOtherWaters" onChange={this.onChange} checked={caughtInOtherWaters === 'Y'} value="Y">{t('ccWhoseWatersWereTheyCaughtInCheckBox3Text')}</Checkbox>
          </MultiChoice>

          {(!this.state.jsEnabled || caughtInOtherWaters === 'Y') &&
            <Fragment>
              <div style={{ borderLeft: 'solid #eee 10px', paddingLeft: '10px' }}>
                <InputField
                  id={'label-otherWaters'}
                  htmlFor={'otherWaters'}
                  input={{
                    autoComplete: 'off',
                    id: 'otherWaters',
                    name: 'otherWaters',
                    value: otherWaters
                  }}
                  meta={{
                    touched: t(otherWatersError),
                    error: t(otherWatersError),
                  }} onChange={this.otherWatersChange}>{t('ccWhoseWatersWereTheyCaughtInInputText')}</InputField>
              </div>
            </Fragment>
          }

          <br />
          <GridRow>
            <SaveAsDraftButton formactionUrl={saveAsDraftFormActionUrl} onClick={this.onSaveAsDraft} />
            <ContinueButton type="submit" id="continue">{t('commonContinueButtonSaveAndContinueButton')}</ContinueButton>
          </GridRow>
        </form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  }

}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    conservation: state.conservation,
    exportPayload: state.exportPayload,
    getAddedLandingsType: state.landingsType,
  };
}

function loadData(store) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }

  return store.dispatch(getConservation(this.documentNumber))
    .then(() => store.dispatch(getExportPayload(this.documentNumber)))
    .then(() => store.dispatch(getLandingType(this.documentNumber)));

}

export const component = connect(
  mapStateToProps,
  {
    addConservation,
    saveConservation,
    getConservation,
    clearConservation,
    dispatchClearErrors,
    getExportPayload,
    getLandingType
  }
)(withTranslation() (WhoseWatersWereTheyCaughtInPage));

WhoseWatersWereTheyCaughtInPage.propTypes = {
  conservation: PropTypes.object.isRequired,
  errors: PropTypes.object,
  exportPayload: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
