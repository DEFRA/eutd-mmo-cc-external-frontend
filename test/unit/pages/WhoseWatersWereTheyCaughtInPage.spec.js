import { mount } from 'enzyme';
import * as React from 'react';
import { component as WhoseWatersWereTheyCaughtInPage} from '../../../src/client/pages/exportCertificates/WhoseWatersWereTheyCaughtInPage';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import { getLandingType } from '../../../src/client/actions/landingsType.actions';
import {  getConservation, getExportPayload, clearConservation, addConservation } from '../../../src/client/actions';
import { render } from '@testing-library/react';
jest.mock('../../../src/client/actions/landingsType.actions');
jest.mock('../../../src/client/actions');

describe('Whose Waters Were They Caught in Page', () => {
  let wrapper;
  let props;

  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res({});
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  props = {
    history: [],
    route: {
      title: 'Create a UK catch certificate for exports',
      previousUri: ':documentNumber/add-landings',
      nextUri: ':documentNumber/how-does-the-export-leave-the-uk',
      path: ':documentNumber/what-export-journey',
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      directLandingUri: ':documentNumber/direct-landing',
      journey: 'catchCertificate',
      progressUri: '/catch-certificates/:documentNumber/progress'
    },
    conservation: {
      caughtInOtherWaters: 'Y',
    },
    match: { params: { documentNumber: 'GBR-2021-CC-D6FBF748C' } }
  };

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDING_TYPE' });
    getConservation.mockReturnValue({ type: 'GET_CONSERVATION' });
    getExportPayload.mockReturnValue({ type: 'GET_EXPORT_PAYLOAD' });
    clearConservation.mockReturnValue({ type: 'CLEAR_CONSERVATION' });
    addConservation.mockReturnValue({ type: 'ADD_CONSERVATION' });
    const store = mockStore({
      errors: {},
      conservation: {},
      landingsType: { landingsEntryOption: 'directLanding' },
      exportPayload: {
        items: [
          {
            product: {
              commodityCode: '03036310',
              presentation: {
                code: 'FIL',
                label: 'Filleted',
              },
              state: {
                code: 'FRO',
                label: 'Frozen',
              },
              species: {
                code: 'COD',
                label: 'Atlantic cod (COD)',
              },
            },
            landings: [
              {
                addMode: false,
                editMode: false,
                model: {
                  id: '99bc2947-c6f4-4012-9653-22dc0b9ad036',
                  vessel: {
                    pln: 'AR190',
                    vesselName: 'SILVER QUEST',
                    homePort: 'TROON AND SALTCOATS',
                    registrationNumber: 'A10726',
                    licenceNumber: '42384',
                    label: 'SILVER QUEST (AR190)',
                  },
                  dateLanded: '2019-01-22',
                  exportWeight: '22',
                },
              },
            ],
          },
        ],
      },
    });
    window.scrollTo = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <WhoseWatersWereTheyCaughtInPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });



  it('should render caught in uk waters check box', () => {
    expect(
      wrapper.find('input[name="caughtInUKWaters"]').exists()
    ).toBeTruthy();
  });

  it('should render caught in eu waters check box', () => {
    expect(
      wrapper.find('input[name="caughtInEUWaters"]').exists()
    ).toBeTruthy();
  });

  it('should render caught in other waters check box', () => {
    expect(
      wrapper.find('input[name="caughtInOtherWaters"]').exists()
    ).toBeTruthy();
  });

  it('should handle submit event', () => {
    wrapper.find('form').simulate('submit', { preventDefault() {} });
  });

  it('should handle on change events setting check box value', () => {
    wrapper
      .find('input[name="caughtInUKWaters"]')
      .simulate('change', { target: { name: 'caughtInUKWaters', value: 'Y' } });
  });

  it('should handle save as draft event', () => {
    wrapper.find('button#saveAsDraft').simulate('click');
  });

  it('should have an id on all inputs', () => {
    expect(wrapper.find('input[id="watersCaughtIn"]').exists()).toBeTruthy();
    expect(wrapper.find('input[id="caughtInEUWaters"]').exists()).toBeTruthy();
    expect(
      wrapper.find('input[id="caughtInOtherWaters"]').exists()
    ).toBeTruthy();
    expect(wrapper.find('input[id="otherWaters"]').exists()).toBeTruthy();
  });

  it('should have a for attribute on all input labels', () => {
    expect(
      wrapper.find('label#label-otherWaters').props()['htmlFor']
    ).toBeDefined();
    expect(wrapper.find('label#label-otherWaters').props()['htmlFor']).toBe(
      'otherWaters'
    );
    expect(
      wrapper.find('label#label-watersCaughtIn').props()['htmlFor']
    ).toBeDefined();
    expect(wrapper.find('label#label-watersCaughtIn').props()['htmlFor']).toBe(
      'watersCaughtIn'
    );
    expect(
      wrapper.find('label#label-caughtInEUWaters').props()['htmlFor']
    ).toBeDefined();
    expect(
      wrapper.find('label#label-caughtInEUWaters').props()['htmlFor']
    ).toBe('caughtInEUWaters');
    expect(
      wrapper.find('label#label-caughtInOtherWaters').props()['htmlFor']
    ).toBeDefined();
    expect(
      wrapper.find('label#label-caughtInOtherWaters').props()['htmlFor']
    ).toBe('caughtInOtherWaters');
  });

  it('should have a back to progress page link', () => {
    expect(wrapper.find('a[href="/catch-certificates/GBR-2021-CC-D6FBF748C/progress"]').exists()).toBeTruthy();
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
     expect(container).toMatchSnapshot();
    });

});


describe('should call history to forbid access', () => {
  it('should not allow access to page when conservation is unauthorised', async () => {
    const mockPush = jest.fn();
    const mockUnauthorised = jest.fn();

    const previousProps = {
      conservation: {
        unauthorised: false,
      }
    };

    let props = {
      match: {
        path: '/',
        url: '/',
        isExact: true,
        params: { documentNumber: '' },
      },
      errors: {},
      route: {
        nextUri: '',
        path: '',
        previousUri: '',
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        title: 'Create a UK catch certificate - GOV.UK',
        progressUri: '/catch-certificates/:documentNumber/progress'
      },
      conservation: {
        unauthorised: true,
      },
      history: {
        push: mockPush,
      },
      unauthorised: mockUnauthorised,
    };

    await new WhoseWatersWereTheyCaughtInPage.WrappedComponent(
      props
    ).componentDidUpdate(previousProps, props);
    expect(mockPush).toHaveBeenCalledWith('/forbidden');
  });

  it('should allow access to page when authorised', async () => {
    const mockPush = jest.fn();
    const mockUnauthorised = jest.fn();

    const previousProps = {
      conservation: {
        unauthorised: false,
      }
    };

    let props = {
      match: {
        path: '/',
        url: '/',
        isExact: true,
        params: { documentNumber: '' },
      },
      errors: {},
      route: {
        nextUri: '',
        path: '',
        previousUri: '',
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        title: 'Create a UK catch certificate - GOV.UK',
        progressUri: '/catch-certificates/:documentNumber/progress'
      },
      conservation: {
        unauthorised: false,
      },
      history: {
        push: mockPush,
      },
      unauthorised: mockUnauthorised,
    };

    await new WhoseWatersWereTheyCaughtInPage.WrappedComponent(
      props
    ).componentDidUpdate(previousProps, props);
    expect(mockPush).not.toHaveBeenCalledWith('/forbidden');
  });
});

describe('Re-direct user to landing entry page when landing entry is null', () => {
  const mockPush = jest.fn();
  const mockGetLandingType = jest.fn();
  const mockGetConservation = jest.fn();
  const mockGetExportPayload = jest.fn();

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should re-direct to landing entry page when landingsEntryOption is null and generatedByContent is false', async () => {
    let props = {
      match: {
        path: '/',
        url: '/',
        isExact: true,
        params: { documentNumber: 'GBR-23423-4234234' },
      },
      errors: {},
      route: {
        nextUri: '',
        path: '',
        landingsEntryUri: ':documentNumber/landings-entry',
        previousUri: '',
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        title: 'Create a UK catch certificate - GOV.UK',
        progressUri: '/catch-certificates/:documentNumber/progress'
      },
      conservation: {},
      history: {
        push: mockPush,
      },
      getAddedLandingsType: {
        landingsEntryOption: null,
        generatedByContent: false,
      },
      getLandingType: mockGetLandingType,
      getConservation: mockGetConservation,
      getExportPayload: mockGetExportPayload,
      exportPayload: {
        items: [
          {
            landings: [],
          },
        ],
      },
    };

    await new WhoseWatersWereTheyCaughtInPage.WrappedComponent(props).componentDidMount();

    expect(mockGetLandingType).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockGetConservation).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockGetExportPayload).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockPush).toHaveBeenCalledWith('GBR-23423-4234234/landings-entry');
  });

  it('should not re-direct to landing entry page when landingsEntryOption is manualEntry and generatedByContent is false', async () => {
    const mockPush = jest.fn();
    const mockGetLandingType = jest.fn();
    const mockGetConservation = jest.fn();
    const mockGetExportPayload = jest.fn();

    let props = {
      match: {
        path: '/',
        url: '/',
        isExact: true,
        params: { documentNumber: 'GBR-23423-4234234' },
      },
      errors: {},
      route: {
        nextUri: '',
        path: '',
        landingsEntryUri: ':documentNumber/landings-entry',
        previousUri: '',
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        title: 'Create a UK catch certificate - GOV.UK',
        progressUri: '/catch-certificates/:documentNumber/progress'
      },
      conservation: {},
      history: {
        push: mockPush,
      },
      getAddedLandingsType: {
        landingsEntryOption: 'manualEntry',
        generatedByContent: false,
      },
      getLandingType: mockGetLandingType,
      getConservation: mockGetConservation,
      getExportPayload: mockGetExportPayload,
      exportPayload: {
        items: [
          {
            product: {
              commodityCode: '03036310',
              presentation: {
                code: 'FIL',
                label: 'Filleted',
              },
              state: {
                code: 'FRO',
                label: 'Frozen',
              },
              species: {
                code: 'COD',
                label: 'Atlantic cod (COD)',
              },
            },
            landings: [
              {
                addMode: false,
                editMode: false,
                model: {
                  id: '99bc2947-c6f4-4012-9653-22dc0b9ad036',
                  vessel: {
                    pln: 'AR190',
                    vesselName: 'SILVER QUEST',
                    homePort: 'TROON AND SALTCOATS',
                    registrationNumber: 'A10726',
                    licenceNumber: '42384',
                    label: 'SILVER QUEST (AR190)',
                  },
                  dateLanded: '2019-01-22',
                  exportWeight: '22',
                },
              },
            ],
          },
        ],
      },
    };

    await new WhoseWatersWereTheyCaughtInPage.WrappedComponent(props).componentDidMount();

    expect(mockGetLandingType).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockGetConservation).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockGetExportPayload).toHaveBeenCalledWith('GBR-23423-4234234');
    expect(mockPush).not.toHaveBeenCalledWith('GBR-23423-4234234/landings-entry');
  });
});

describe('Disable backLink when there is no product details for items returned', () => {
    let wrapper;
    let props;

    const mockStore = configureStore([
      thunk.withExtraArgument({
        orchestrationApi: {
          get: () => {
            return new Promise((res) => {
              res({});
            });
          },
          post: () => {
            return new Promise((res) => {
              res({});
            });
          },
        },
      }),
    ]);

    props = {
      history: [],
      route: {
        title: 'Create a UK catch certificate for exports',
        previousUri: ':documentNumber/add-landings',
        nextUri: ':documentNumber/how-does-the-export-leave-the-uk',
        path: ':documentNumber/what-export-journey',
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        directLandingUri: ':documentNumber/direct-landing',
        journey: 'catchCertificate',
        progressUri: '/catch-certificates/:documentNumber/progress'
      },
      conservation: {
        caughtInOtherWaters: 'Y',
      },
      match: { params: { documentNumber: 'GBR-2021-CC-D6FBF748C' } }
    };

    beforeEach(() => {
      getLandingType.mockReturnValue({ type: 'GET_LANDING_TYPE' });
      getConservation.mockReturnValue({ type: 'GET_CONSERVATION' });
      getExportPayload.mockReturnValue({ type: 'GET_EXPORT_PAYLOAD' });
      clearConservation.mockReturnValue({ type: 'CLEAR_CONSERVATION' });
      addConservation.mockReturnValue({ type: 'ADD_CONSERVATION' });
    });

  it('should not render backLink when there are no product details  exists', () => {

    const store = mockStore({
      errors: {},
      conservation: {},
      landingsType: { landingsEntryOption: 'directLanding' },
      exportPayload: {
        items: [],
      },
    });
    window.scrollTo = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <WhoseWatersWereTheyCaughtInPage {...props} />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('BackLink').exists()).toBeFalsy();
  });

  it('should render backLink when there is a product details exsists', () => {

    const store = mockStore({
      errors: {},
      conservation: {},
      landingsType: { landingsEntryOption: 'directLanding' },
      exportPayload: {
        items: [{
          product: {
            commodityCode: '03036310',
            presentation: {
              code: 'FIL',
              label: 'Filleted',
            },
            state: {
              code: 'FRO',
              label: 'Frozen',
            },
            species: {
              code: 'COD',
              label: 'Atlantic cod (COD)',
            },
          },
          landings: [
            {
              addMode: false,
              editMode: false,
              model: {
                id: '99bc2947-c6f4-4012-9653-22dc0b9ad036',
                vessel: {
                  pln: 'AR190',
                  vesselName: 'SILVER QUEST',
                  homePort: 'TROON AND SALTCOATS',
                  registrationNumber: 'A10726',
                  licenceNumber: '42384',
                  label: 'SILVER QUEST (AR190)',
                },
                dateLanded: '2019-01-22',
                exportWeight: '22',
              },
            },
          ],
        }],
      },
    });
    window.scrollTo = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <WhoseWatersWereTheyCaughtInPage {...props} />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('BackLink').exists()).toBeTruthy();
    expect(wrapper.find('BackLink').props().href).toContain('GBR-2021-CC-D6FBF748C/direct-landing');
  });
});