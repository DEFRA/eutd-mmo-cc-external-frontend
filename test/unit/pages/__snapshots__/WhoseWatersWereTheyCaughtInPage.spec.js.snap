// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whose Waters Were They Caught in Page should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="GBR-2021-CC-D6FBF748C/direct-landing"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Whose waters were the fish or shellfish caught in?
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/conservation"
        method="POST"
      >
        <input
          name="currentUri"
          type="hidden"
          value=":documentNumber/what-export-journey"
        />
        <input
          name="nextUri"
          type="hidden"
          value=":documentNumber/how-does-the-export-leave-the-uk"
        />
        <input
          name="dashboardUri"
          type="hidden"
          value="/create-catch-certificate/catch-certificates"
        />
        <input
          name="journey"
          type="hidden"
          value="catchCertificate"
        />
        <div
          style="display: none;"
        >
          <label
            class="e1vcssft0 css-mpbcqr-StyledHoc"
            id="label-undefined"
          >
            <input
              class="css-17ieumk"
              type="checkbox"
            />
            <span
              class="css-1ogj8nh"
            />
          </label>
        </div>
        <div
          class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
        >
          <span
            class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
          >
            Select all that apply
          </span>
          
          <label
            class="e1vcssft0 css-mpbcqr-StyledHoc"
            for="watersCaughtIn"
            id="label-watersCaughtIn"
          >
            <input
              class="css-17ieumk"
              id="watersCaughtIn"
              name="caughtInUKWaters"
              type="checkbox"
              value="Y"
            />
            <span
              class="css-1ogj8nh"
            >
              UK, British Isles
            </span>
          </label>
          <label
            class="e1vcssft0 css-mpbcqr-StyledHoc"
            for="caughtInEUWaters"
            id="label-caughtInEUWaters"
          >
            <input
              class="css-17ieumk"
              id="caughtInEUWaters"
              name="caughtInEUWaters"
              type="checkbox"
              value="Y"
            />
            <span
              class="css-1ogj8nh"
            >
              EU
            </span>
          </label>
          <label
            class="e1vcssft0 css-mpbcqr-StyledHoc"
            for="caughtInOtherWaters"
            id="label-caughtInOtherWaters"
          >
            <input
              class="css-17ieumk"
              id="caughtInOtherWaters"
              name="caughtInOtherWaters"
              type="checkbox"
              value="Y"
            />
            <span
              class="css-1ogj8nh"
            >
              Other
            </span>
          </label>
        </div>
        <div
          style="border-left: 10px solid #eee; padding-left: 10px;"
        >
          <label
            class="e1vcssft0 e1vcssft0 css-5tg97n-StyledLabel-StyledHoc-StyledHoc e4u1u2g0"
            for="otherWaters"
            id="label-otherWaters"
          >
            <span
              class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
            >
              Whose waters?
            </span>
            
            <input
              autocomplete="off"
              class="e1vcssft0 css-4n7f6s-StyledInput-StyledHoc e4u1u2g0"
              id="otherWaters"
              name="otherWaters"
              type="text"
              value=""
            />
          </label>
        </div>
        <br />
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/conservation/saveAsDraft"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/catch-certificates/GBR-2021-CC-D6FBF748C/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
